### PR TITLE
Bug fix: PubSub tests "overrun" their timers (Unity)

### PR DESCRIPTION
**Pull request by JTonoJimenez**

### CODE REVIEW

File: targets/unity-v2/Testing/Tests/Shared/Uunit/UUnitTestContext.cs [hunk 1]
```diff
@@ -44,6 +44,10 @@ public UUnitTestContext(UUnitTestCase testInstance, Action<UUnitTestContext> tes
 
         public void EndTest(UUnitFinishState finishState, string resultMsg)
         {
+            if (finishState != UUnitFinishState.PENDING)
+            {
+                return;
```
Comment (pgilmorepf) [inline targets/unity-v2/Testing/Tests/Shared/Uunit/UUnitTestContext.cs]: This is not sufficient.
This just hides the problem.

There are state transitions that are normal/acceptable:
PENDING->PASSED
PENDING->FAILED
PASSED->FAILED

There are state transitions that should be explicitly disallowed, and attempting them should result in failure:
FAILED->PASSED (Keep original fail message, append "You can't mark a test as finished twice")
PASSED->PASSED (Fail with a new message that says "You can't mark a test as finished twice")
FAILED->FAILED (Keep/merge both messages)

**Comment by MichaelMackendMS** [comment]
Can we get a bit of description as to how these changes addresses the problem?  Like: 
1) What caused the problem
2) How this addresses it

It would save a lot of time in trying to reverse engineer what the reasoning was and why we should feel like this is good to go. 

Thanks!

**Comment by Vicodex** [comment]
retest

**Comment by Vicodex** [comment]
retest

File: targets/unity-v2/Testing/Tests/Shared/Uunit/UUnitTestContext.cs [hunk 2]
```diff
@@ -44,6 +44,29 @@ public UUnitTestContext(UUnitTestCase testInstance, Action<UUnitTestContext> tes
...
+                else
+                {
+                    if (finishState != UUnitFinishState.FAILED)
+                    {
+                        resultMsg += "You can't mark a test as finished twice.\n";
+                    }
+                    TestResultMsg = resultMsg;
+                    return;
+                }
+            }
```
Comment (pgilmorepf) [inline targets/unity-v2/Testing/Tests/Shared/Uunit/UUnitTestContext.cs]: It's really hard to follow this logic.

I would prefer:

if (condition && condition)
// logic
else if (condition && condition)
// logic
else if (condition && condition)
// logic
else if (condition
// logic
else
// logic

Basically, your logic above may be right, but if exactly one state is wrong, then we have to reinvent this whole code. The switch statement I describe above may be more code, but it's a lot easier to debug and add a condition pair that may not be covered by your logic.

**Comment by pgilmorepf** [comment]
retest

**Comment by pgilmorepf** [comment]
retest

**Pull request merged**